luomu-tpacketv3/src/ringbuf.rs

```diff
@@ -0,0 +1,160 @@
+use crate::if_packet;
+use std::fmt;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+/// Descriptor for single packet within a block in mmapped ringbuffer
+pub struct PacketDescriptor<'a> {
+    ptr: *mut u8,
+    hdr: &'a if_packet::tpacket3_hdr,
+}
+
+impl<'a> PacketDescriptor<'a> {
+    #[allow(clippy::cast_ptr_alignment)]
+    pub fn get_next(self) -> PacketDescriptor<'a> {
+        let next_ptr = unsafe { self.ptr.offset(self.hdr.tp_next_offset as isize) };
+        let next_hdr_ptr = next_ptr as *const if_packet::tpacket3_hdr;
+        PacketDescriptor {
+            ptr: next_ptr,
+            hdr: unsafe { next_hdr_ptr.as_ref().unwrap() },
+        }
+    }
+
+    pub fn get_timestamp(&self) -> SystemTime {
+        UNIX_EPOCH + Duration::new(self.hdr.tp_sec as u64, self.hdr.tp_nsec)
+    }
+
+    /// Copies packet data from buffer into a vector.
+    pub fn get_packet_data(&self) -> &'a [u8] {
+        let data_ptr = unsafe { self.ptr.offset(self.hdr.tp_mac as isize) };
+        unsafe { std::slice::from_raw_parts(data_ptr, self.hdr.tp_snaplen as usize) }
+    }
+}
+
+impl<'a> From<*mut u8> for PacketDescriptor<'a> {
+    #[allow(clippy::cast_ptr_alignment)]
+    fn from(ptr: *mut u8) -> Self {
+        let hdr_ptr = ptr as *const if_packet::tpacket3_hdr;
+        let hdr = unsafe { hdr_ptr.as_ref().unwrap() };
+        PacketDescriptor { ptr, hdr }
+    }
+}
+
+#[derive(Debug)]
+/// Descriptor for a block inside the mmapped ringbuffer.
+/// Block is either owned by kernel or ready. Ready block contains 1 or more
+/// packets which can be read. Once all packets have been consumed,
+/// flush() needs to be called to indicate kernel that it free to use this
+/// block again
+pub struct BlockDescriptor<'a> {
+    ptr: *mut u8,                                // pointer to the start of the data
+    desc: &'a mut if_packet::tpacket_block_desc, // the actual block descriptor
+    name: String,
```

Comment:
Is this `name` unnecessary? It can be constructed in `Display` trait implementation when needed.

Reply:
Removed.